include/ibamr/BrinkmanPenalizationAdvDiff.h

```diff
@@ -0,0 +1,306 @@
+     */
+    ~BrinkmanPenalizationAdvDiff() = default;
+
+    /*!
+     * \brief Set the time interval in which Brinkman forcing is computed.
+     */
+    void setTimeInterval(double current_time, double new_time);
+
+    /*!
+     * \brief Preprocess routine before computing Brinkman penalization terms.
+     *
+     */
+    void preprocessBrinkmanPenalizationAdvDiff(double current_time, double new_time, int num_cycles);
+
+    /*!
+     * \brief Postprocess routine after computing Brinkman penalization terms.
+     *
+     */
+    void postprocessBrinkmanPenalizationAdvDiff(double current_time, double new_time, int num_cycles);
+
+    /*!
+     * \brief Set Brinkman penalization penalty factor for all level sets.
+     */
+    void setPenaltyCoefficient(double eta_penalty_coef);
+
+    /*!
+     * \brief Set the number of interface cells for all level sets.
+     */
+    void setNumInterfaceCells(double num_interface_cells);
+
+    /*!
+     * \brief Get the name of the object.
+     */
+    const std::string& getName() const
+    {
+        return d_object_name;
+    } // getName
+
+    /*!
+     * \brief Get the current time interval \f$ [t^{n+1}, t^n] \f$ in which Brinkman
+     * velocity is computed.
+     */
+    std::pair<double, double> getCurrentTimeInterval() const
+    {
+        return std::make_pair(d_new_time, d_current_time);
+    } // getCurrentTimeInterval
+
+    /*!
+     * \brief Function to determine if a transported quantity has Brinkman
+     * boundary conditions associated with it.
+     */
+    bool hasBrinkmanBoundaryCondition(SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var) const
+    {
+        return d_Q_bc.find(Q_var) != d_Q_bc.end();
+    } // hasBrinkmanBoundaryCondition
+
+    /*!
+     * \brief Function specifying the optional forcing function for inhomogeneous Neumann boundary conditions \f$ n \dot
+     * \Q = \gamma \f$
+     *
+     * The user must set the patch data B_idx such that \f$ n \dot B = \gamma \f$.
+     * The applied forcing term is then computed internally as \f$ \nabla \dot (\chi B) - \chi \nabla \dot B \f$.
+     * Note that B_idx contains side-centered patch data.
+     */
+    using BrinkmanInhomogeneousNeumannBCsFcnPtr = void (*)(int B_idx,
+                                                           SAMRAI::tbox::Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                                           double time,
+                                                           void* ctx);
+
+    /*!
+     * \brief Register a transported quantity with this object, along with the solid level set
+     * variable for which to apply a specified boundary condition.
+     *
+     * \note This function can only be used to register homogeneous/inhomogeneous Dirichlet BCs, and homogeneous Neumann
+     * BCs.
+     */
+    void
+    registerDirichletHomogeneousNeumannBC(SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var,
```

Comment:
Could we use the existing BC infrastructure? I would like to avoid propagating different ways of setting BCs if possible

Reply:
(Or is this doing a different kind of BC setting?)

Reply:
This is imposing BCs on embedded interfaces and *not* on the domain boundaries; so it is an entirely different thing. We do provide a description in the class header.

Reply:
Perhaps the comment here (and in related functions) could state that it is specifying the BC along the interface determined by the level set variable.

Also, instead of having separate functions for Dirichlet, Neumann, and homogeneous/inhomogenous, how about just one function that takes in a `bc_type` `enum` and the callback, but if the callback is `NULL` treat as homogeneous?

Reply:
> > A couple of comments:
> > 
> > 1. It seems like Brinkmann penalization is only implemented for `AdvDiffSemiImplicitHierarchyIntegrator`. If that is right, then I think that either the other implementations need to generate an error message if Brinkmann penalization is set. Alternatively, only `AdvDiffSemiImplicitHierarchyIntegrator` should know about Brinkmann at this time. (@drwells do you have an opinion?)
> 
> This class needs to solve diffusion equations implicitly, hence `AdvDiffSemiImplicitHierarchyIntegrator`.

Sure, that is fine --- but in that case, I am asking why it makes sense for `AdvDiffHierarchyIntegrator` to know about the Brinkmann object. It seems more natural just to have the `AdvDiffSemiImplicitHierarchyIntegrator` know about the Brinkmann object.

If you want `AdvDiffHierarchyIntegrator` to know about the Brinkmann object, then it seems like the default implementation should be that it emits an error message if you try to use it with something other than `AdvDiffSemiImplicitHierarchyIntegrator`.